src/org/sosy_lab/java_smt/solvers/princess/PrincessModel.java

```diff
@@ -237,29 +244,72 @@ public String toString() {
     return model.toString();
   }
 
-  @Override
-  public void close() {}
+  /** Tries to determine the Sort of a Term. */
+  private Sort getSort(IExpression pTerm) {
+    // Just using sortof() won't work as Princess may have simplified the original term
+    // FIXME: This may also affect other parts of the code that use sortof()
+    if (pTerm instanceof ITimes) {
+      ITimes times = (ITimes) pTerm;
+      return getSort(times.subterm());
+    } else if (pTerm instanceof IPlus) {
+      IPlus plus = (IPlus) pTerm;
+      return getSort(plus.apply(0));
+    } else if (pTerm instanceof IFormula) {
+      return creator.getBoolType();
+    } else {
+      // TODO: Do we need more cases?
+      return Sort$.MODULE$.sortOf((ITerm) pTerm);
+    }
+  }
 
-  @Override
-  protected IExpression evalImpl(IExpression formula) {
-    IExpression evaluation = evaluate(formula);
-    if (evaluation == null) {
-      // fallback: try to simplify the query and evaluate again.
-      evaluation = evaluate(creator.getEnv().simplify(formula));
+  /**
+   * Simplify rational values.
+   *
+   * <p>Rewrite <code>a/1</code> as <code>a</code>, otherwise return the term unchanged
+   */
+  private ITerm simplifyRational(ITerm pTerm) {
+    // TODO: Reduce the term further?
+    // TODO: Also do this for the values in the model?
+    if (Sort$.MODULE$.sortOf(pTerm).equals(creator.getRationalType())
+        && pTerm instanceof IFunApp
+        && ((IFunApp) pTerm).fun().name().equals("Rat_frac")
+        && pTerm.apply(1).equals(new IIntLit(IdealInt.ONE()))) {
+      return Rationals.int2ring((ITerm) pTerm.apply(0));
     }
-    return evaluation;
+    return pTerm;
   }
 
-  @Nullable
-  private IExpression evaluate(IExpression formula) {
-    if (formula instanceof ITerm) {
-      Option<ITerm> out = model.evalToTerm((ITerm) formula);
-      return out.isEmpty() ? null : out.get();
-    } else if (formula instanceof IFormula) {
-      Option<IExpression> out = model.evalExpression(formula);
-      return out.isEmpty() ? null : out.get();
+  @Override
+  protected @Nullable IExpression evalImpl(IExpression expr) {
+    // Extending the partial model does not seem to work in Princess if the formula uses rational
+    // variables. To work around this issue we (temporarily) add the formula to the assertion
+    // stack and then repeat the sat-check to get the value.
+    if (expr instanceof ITerm) {
```

Comment:
How expensive is this workaround? 

Can we ask the Princess developers for a proper evaluation? I would assume that Princess could implement this internally, including caching for repeated calls, memoization of assignments comparing across different evaluations, etc.

Reply:
> How expensive is this workaround?

I've just run bmc on the `sv-comp` task set as a benchmark:

![Screenshot From 2024-11-17 17-31-55](https://github.com/user-attachments/assets/5c938246-cb33-400a-9990-7e8a7245d4dc)
The brown line uses Princess with the `main`, and the green(?) line is for the updated version from this branch. Since we're not using floating point or string formulas, most of the difference between the two should be down to the changes in `evalImpl` (and possibly some changes in Princess itself). Note that the old version (from `main`) actually crashes on quite a number of the tasks due to https://github.com/sosy-lab/java-smt/issues/308. The workaround has the side-effect of also fixing this issue. If we filter for just the tasks that return the correct result with the Princess version from `main` the graph looks like this:

![Screenshot From 2024-11-17 17-32-45](https://github.com/user-attachments/assets/28eb7f0f-f6b0-4f21-872a-3e2cba8915f8)
From those results it doesn't look like the work-around is adding too much overhead, especially as the problem size increases. It might still be worth looking into other benchmarks that don't use `bmc` and see if there is any significant difference there.

> Can we ask the Princess developers for a proper evaluation? I would assume that Princess could implement this internally, including caching for repeated calls, memoization of assignments comparing across different evaluations, etc.

That would of course be ideal. We already have two Princess tickets (https://github.com/uuverifiers/princess/issues/7 and https://github.com/uuverifiers/princess/issues/8) about issues with the evaluation of rational formulas, and another one (https://github.com/uuverifiers/princess/issues/17) for the crashes on integer formulas that I mentioned earlier (see #308). The first two are already quite old and I don't know what the current status is. Maybe we could just ask for an update? For the last one I was told that they will need more time for a fix as the issues doesn't seem to be entirely trivial. Because of that I would prefer to merge this branch with the workaround now and then come back later once the issues have been fixed in Princess and remove the workaround again.